### PR TITLE
Admin user management/ adding users as admins for multiple schools

### PR DESCRIPTION
**Pull request by eadams17**
## WHAT
add ability to add users as admins for multiple schools via the internal tool and suppress emails for existing admin users

## WHY
we want to have this functionality for the internal tool and we don't want to send more than one email once a user has been made an admin (see Notion card discussion)

## HOW
update admin check in `CMS::SchoolAdminsController` to allow admin creation for additional schools as well as add checks for multiple `SchoolsAdmins` records in the workers to only send an email on the first admin addition

### Screenshots
(If applicable. Also, please censor any sensitive data)

### Notion Card Links
(Please provide links to any relevant Notion card(s) relevant to this PR.)
https://www.notion.so/quill/Error-when-adding-an-admin-to-multiple-schools-in-Quill-CMS-6029b30dcee041e7b6208b257a83f154

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  yes
Have you deployed to Staging? | yes
Self-Review: Have you done an initial self-review of the code below on Github? | yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | yes

### CODE REVIEW

File: services/QuillLMS/app/controllers/admins_controller.rb [hunk 1]
```diff
@@ -147,8 +147,9 @@ def create_and_link_accounts
     if SchoolsAdmins.exists?(user: @teacher, school: @school)
       @message = t('admin_created_account.existing_account.admin.linked', school_name: @school.name)
     else
+      existing_admin = SchoolsAdmins.find_by(user_id: @teacher.id)
```
Comment (brendanshean) [inline services/QuillLMS/app/controllers/admins_controller.rb:150]: I would use a `exists?` here since we aren't using the record at all, but merely just it's existence:

`SchoolsAdmins.exists?(user_id: @teacher.id)`

The `find_by` usage [here](https://github.com/empirical-org/Empirical-Core/blob/d99add1f70ccf26434fe0a0410fc745ab37779bb/services/QuillLMS/app/controllers/cms/school_admins_controller.rb#L63) is appropriate since we're using the record's values.

File: services/QuillLMS/app/workers/admin_dashboard/made_school_admin_change_school_email_worker.rb [hunk 2]
```diff
@@ -11,7 +11,9 @@ def perform(user_id, admin_user_id, new_school_id, existing_school_id)
 
     return unless user && admin_name && new_school && existing_school
 
-    user.mailer_user.send_admin_dashboard_made_school_admin_change_school_email(admin_name, new_school, existing_school)
+    if SchoolsAdmins.where(user_id: user_id).count == 1
```
Comment (brendanshean) [inline services/QuillLMS/app/workers/admin_dashboard/made_school_admin_change_school_email_worker.rb]: This conditional clause violates the law of demeter.  We can fix that by adding a boolean method to user.rb

```
def is_admin_for_one_school?
  schools_admins.count == 1
end
```
then the conditional becomes:

`if user.is_admin_for_one_school?`

File: services/QuillLMS/app/workers/admin_dashboard/made_school_admin_email_worker.rb [hunk 3]
```diff
@@ -10,7 +10,9 @@ def perform(user_id, admin_user_id, school_id)
 
     return unless user && admin_name && school_name
 
-    user.mailer_user.send_admin_dashboard_made_school_admin_email(admin_name, school_name)
+    if SchoolsAdmins.where(user_id: user_id).count == 1
```
Comment (brendanshean) [inline services/QuillLMS/app/workers/admin_dashboard/made_school_admin_email_worker.rb]: use `user.is_admin_for_one_school?` mentioned earlier

File: services/QuillLMS/app/workers/admin_dashboard/made_school_admin_link_school_email_worker.rb [hunk 4]
```diff
@@ -10,7 +10,9 @@ def perform(user_id, admin_user_id, school_id)
 
     return unless user && admin_name && school
 
-    user.mailer_user.send_admin_dashboard_made_school_admin_link_school_email(admin_name, school)
+    if SchoolsAdmins.where(user_id: user_id).count == 1
```
Comment (brendanshean) [inline services/QuillLMS/app/workers/admin_dashboard/made_school_admin_link_school_email_worker.rb]: use `user.is_admin_for_one_school?` mentioned earlier

File: services/QuillLMS/app/workers/internal_tool/made_school_admin_change_school_email_worker.rb [hunk 5]
```diff
@@ -10,7 +10,9 @@ def perform(user_id, new_school_id, existing_school_id)
 
     return unless user && new_school && existing_school
 
-    user.mailer_user.send_internal_tool_made_school_admin_change_school_email(new_school, existing_school)
+    if SchoolsAdmins.where(user_id: user_id).count == 1
```
Comment (brendanshean) [inline services/QuillLMS/app/workers/internal_tool/made_school_admin_change_school_email_worker.rb]: use `user.is_admin_for_one_school?` mentioned earlier

File: services/QuillLMS/app/workers/internal_tool/made_school_admin_email_worker.rb [hunk 6]
```diff
@@ -9,7 +9,9 @@ def perform(user_id, school_id)
 
     return unless user && school_name
 
-    user.mailer_user.send_internal_tool_made_school_admin_email(school_name)
+    if SchoolsAdmins.where(user_id: user_id).count == 1
```
Comment (brendanshean) [inline services/QuillLMS/app/workers/internal_tool/made_school_admin_email_worker.rb]: use `user.is_admin_for_one_school?` mentioned earlier

File: services/QuillLMS/app/workers/internal_tool/made_school_admin_link_school_email_worker.rb [hunk 7]
```diff
@@ -9,7 +9,9 @@ def perform(user_id, school_id)
 
     return unless user && school
 
-    user.mailer_user.send_internal_tool_made_school_admin_link_school_email(school)
+    if SchoolsAdmins.where(user_id: user_id).count == 1
```
Comment (brendanshean) [inline services/QuillLMS/app/workers/internal_tool/made_school_admin_link_school_email_worker.rb]: use `user.is_admin_for_one_school?` mentioned earlier

File: services/QuillLMS/app/controllers/cms/school_admins_controller.rb [hunk 8]
```diff
@@ -42,7 +42,13 @@ def set_school
 
   private def handle_school_admin_save(user, school_id, new_user)
     determine_school_admin_worker(user, school_id, new_user)
-    returned_message = new_user ? t('admin.make_admin') : t('admin_created_account.existing_account.admin.new')
+    returned_message = t('admin_created_account.existing_account.admin.new')
```
Comment (brendanshean) [inline services/QuillLMS/app/controllers/cms/school_admins_controller.rb]: This particular method has become complicated by virtue of the returned_message logic.  I would break out a new method:


```
private def handle_school_admin_save(user, school_id, new_user)
    determine_school_admin_worker(user, school_id, new_user)
    
    if params[:is_make_admin_button]
      flash[:success] = t('admin.make_admin')
      redirect_to cms_school_path(school_id)
    else
      render json: { message: school_admin_save_message(new_user, user) }, status: 200
    end
  end

  private def school_admin_save_message(new_user, user)
    if new_user
      t('admin.make_admin')
    elsif user.is_admin_to_multiple_schools?  # note that this is a new method to write for user.rb
      t('admin_created_account.existing_account.admin.admin_for_other_school')
    else
      t('admin_created_account.existing_account.admin.new')
    end
  end
```

File: services/QuillLMS/app/controllers/cms/school_admins_controller.rb [hunk 9]
```diff
@@ -54,18 +60,19 @@ def set_school
 
   private def create_admin_user_for_existing_user(user, new_user: false)
     school_id = params[:school_id]
+    school_admin_for_school = SchoolsAdmins.find_by(school_id: school_id, user_id: user.id)
```
Comment (brendanshean) [inline services/QuillLMS/app/controllers/cms/school_admins_controller.rb]: Since the school name is what we're really after here, what about going straight to that:

```
school_name = SchoolsAdmins.find_by(school_id: school_id, user_id: user.id)&.school&.name

if school_name
  return render json: { message: t('admin_created_account.existing_account.admin.linked', school_name: school_name) }
end
```

**Review by brendanshean — APPROVED** [review]
Some refactoring comments, otherwise looks good to me!

File: services/QuillLMS/app/controllers/cms/school_admins_controller.rb (hunk 8 above)
Comment (brendanshean) [reply services/QuillLMS/app/controllers/cms/school_admins_controller.rb]: See [comment](https://github.com/empirical-org/Empirical-Core/pull/10211#discussion_r1089299382) for note on `user.is_admin_to_multiple_schools?`

**Review by emilia-friedberg — APPROVED** [review]
I agree with Brendan's comments, and also wonder if it is worth adding a third test case to those email worker tests (we're checking for `nil` and `not nil` for the user, but not the case where there is a user but we don't want to send an email). Otherwise LGTM!

File: services/QuillLMS/app/workers/admin_dashboard/made_school_admin_change_school_email_worker.rb (hunk 2 above)
Comment (eadams17) [reply services/QuillLMS/app/workers/admin_dashboard/made_school_admin_change_school_email_worker.rb]: Cool, I'll make that change!

File: services/QuillLMS/app/controllers/cms/school_admins_controller.rb (hunk 8 above)
Comment (eadams17) [reply services/QuillLMS/app/controllers/cms/school_admins_controller.rb]: yeah, that seems cleaner-- thanks for the suggestion!
